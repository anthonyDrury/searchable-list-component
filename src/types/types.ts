export type Contact = {
  id: string;
  name: string;
  attended?: boolean;
  displayImageURL?: string;
  email?: string;
};

export type SelectedContactsRecord = Record<Contact["id"], boolean>;
