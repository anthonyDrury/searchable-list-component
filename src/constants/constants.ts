export const DEFAULT_CONTACT_IMAGE_URL = "/default_contact_display_image.jpg";

export const TEXT_BLACK = "#20374B";
export const TEXT_LIGHT = "#5A6D80";
export const TEXT_PRIMARY = "#6B46EF";

export const CONTENT_BACKGROUND = "#FFFFFF";
export const CONTENT_BACKGROUND_PRIMARY = "#F4F1FD";
export const CONTENT_BACKGROUND_SECONDARY = "#F2F5F7";

export const BORDER_COLOR = "#E4E5E8";
