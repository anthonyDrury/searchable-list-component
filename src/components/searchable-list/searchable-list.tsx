import { useMemo, useState } from "react";
import styled from "styled-components";
import { CONTENT_BACKGROUND } from "../../constants/constants";
import { SelectedContactsProvider } from "../../contexts/selected-contacts-context/selected-contacts.context";
import { Contact, SelectedContactsRecord } from "../../types/types";
import { ContactList } from "../contact-list/contact-list";
import { SearchField } from "../search-field/search-field";
import { SectionHeader } from "../section-header/section-header";

type SearchableListProps = {
  onSelectedContactsChange: (
    newSelectedContacts: SelectedContactsRecord
  ) => void;
  displayEmail?: boolean;
  contacts: Contact[];
};
export const SearchableList = ({
  onSelectedContactsChange,
  displayEmail = false,
  contacts,
}: SearchableListProps) => {
  const [searchTerm, setSearchTerm] = useState("");

  const [attendedContacts, absentContacts] = useMemo(
    () => filterContacts(contacts, searchTerm),
    [contacts, searchTerm]
  );

  return (
    <SearchableListContainer>
      <SearchField onSearch={setSearchTerm} />
      <SelectedContactsProvider
        onSelectedContactsChange={onSelectedContactsChange}
      >
        <SectionHeader headerName="Attended">
          <ContactList
            contacts={attendedContacts}
            displayEmail={displayEmail}
            enabled={false}
          />
        </SectionHeader>
        <SectionHeader headerName="Absent">
          <ContactList
            contacts={absentContacts}
            displayEmail={displayEmail}
            enabled={false}
          />
        </SectionHeader>
      </SelectedContactsProvider>
    </SearchableListContainer>
  );
};

const filterContacts = (
  contacts: Contact[],
  searchTerm: string
): [Contact[], Contact[]] =>
  contacts.reduce(([attended, absent], cur) => {
    const filterMatch = searchTerm
      ? cur.name?.toLowerCase().includes(searchTerm.toLowerCase())
      : true;

    if (filterMatch) {
      if (cur.attended) {
        return [[...attended, cur], absent];
      } else {
        return [attended, [...absent, cur]];
      }
    }
    return [attended, absent];
  }, EMPTY_CONTACTS);

// Outside of the component to ensure no empty value changes on re-renders
const EMPTY_CONTACTS: [Contact[], Contact[]] = [[], []];

const SearchableListContainer = styled.div`
  background-color: ${CONTENT_BACKGROUND};
  max-width: 400px;
`;
