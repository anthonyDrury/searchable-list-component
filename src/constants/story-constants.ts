import { Contact } from "../types/types";
import { DEFAULT_CONTACT_IMAGE_URL } from "./constants";

export const DEFAULT_CONTACT: Contact = {
  id: "00001",
  name: "John Doe",
  displayImageURL: DEFAULT_CONTACT_IMAGE_URL,
  email: "test@test.com",
  attended: false,
};

export const DEFAULT_CONTACT_LIST: Contact[] = [
  {
    id: "00000",
    name: "John Doe",
    displayImageURL: "/DISPLAY_IMAGE_0.png",
    email: "",
    attended: false,
  },
  {
    id: "00001",
    name: "Jane Doe",
    displayImageURL: "/DISPLAY_IMAGE_1.png",
    email: "test@test.com",
    attended: false,
  },
  {
    id: "00002",
    name: "Alice Smith",
    displayImageURL: "/DISPLAY_IMAGE_2.png",
    email: "alice@example.com",
    attended: true,
  },
  {
    id: "00003",
    name: "Bob Johnson",
    displayImageURL: "/DISPLAY_IMAGE_3.png",
    email: "bob@example.com",
    attended: false,
  },
  {
    id: "00004",
    name: "Charlie Brown",
    displayImageURL: "/DISPLAY_IMAGE_4.png",
    email: "charlie@example.com",
    attended: true,
  },
  {
    id: "00005",
    name: "Diana Prince",
    displayImageURL: "/DISPLAY_IMAGE_5.png",
    email: "diana@example.com",
    attended: false,
  },
  {
    id: "00006",
    name: "Eve Adams",
    displayImageURL: "/DISPLAY_IMAGE_6.png",
    email: "eve@example.com",
    attended: true,
  },
  {
    id: "00007",
    name: "Frank Castle",
    displayImageURL: "/DISPLAY_IMAGE_7.png",
    email: "frank@example.com",
    attended: false,
  },
  {
    id: "00008",
    name: "Grace Hopper",
    displayImageURL: "/DISPLAY_IMAGE_8.png",
    email: "grace@example.com",
    attended: true,
  },
  {
    id: "00009",
    name: "Hank Pym",
    displayImageURL: "/DISPLAY_IMAGE_9.png",
    email: "hank@example.com",
    attended: false,
  },
  {
    id: "00010",
    name: "Ivy Wayne",
    displayImageURL: undefined,
    email: undefined,
    attended: false,
  },
];
