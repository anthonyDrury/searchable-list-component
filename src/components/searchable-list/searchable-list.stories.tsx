import type { Meta, StoryObj } from "@storybook/react";
import { DEFAULT_CONTACT_LIST } from "../../constants/story-constants";
import { SearchableList } from "./searchable-list";

import { fn } from "@storybook/test";

const meta = {
  title: "Example/SearchableList",
  component: SearchableList,
  parameters: {},
  tags: ["autodocs"],
  argTypes: {
    displayEmail: { control: "boolean" },
    contacts: {
      control: {
        type: "object",
      },
    },
  },
  args: {
    onSelectedContactsChange: fn(),
    contacts: DEFAULT_CONTACT_LIST,
  },
} satisfies Meta<typeof SearchableList>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    onSelectedContactsChange: fn(),
    contacts: DEFAULT_CONTACT_LIST,
  },
};
export const DisplayEmail: Story = {
  args: {
    onSelectedContactsChange: fn(),
    contacts: DEFAULT_CONTACT_LIST,
    displayEmail: true,
  },
};
