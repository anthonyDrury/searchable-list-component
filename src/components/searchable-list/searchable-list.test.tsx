import React, { act } from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { SearchableList } from "./searchable-list";
import { DEFAULT_CONTACT_LIST } from "../../constants/story-constants";
import { Contact, SelectedContactsRecord } from "../../types/types";
import {
  CONTENT_BACKGROUND_PRIMARY,
  DEFAULT_CONTACT_IMAGE_URL,
} from "../../constants/constants";

describe("SearchableList", () => {
  const onSelectedContactsChange = jest.fn();

  it("Matches the default snapshot", () => {
    const { container } = render(
      <SearchableList
        contacts={DEFAULT_CONTACT_LIST}
        onSelectedContactsChange={onSelectedContactsChange}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("Matches the email display snapshot", () => {
    const { container } = render(
      <SearchableList
        contacts={DEFAULT_CONTACT_LIST}
        onSelectedContactsChange={onSelectedContactsChange}
        displayEmail
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("Renders the Searchable lists and all details", () => {
    render(
      <SearchableList
        contacts={DEFAULT_CONTACT_LIST}
        onSelectedContactsChange={onSelectedContactsChange}
      />
    );

    const searchInput = screen.getByPlaceholderText("Search");
    expect(searchInput).toBeInTheDocument();
    const attendedHeader = screen.getByText("Attended");
    expect(attendedHeader).toBeInTheDocument();
    const absentHeader = screen.getByText("Absent");
    expect(absentHeader).toBeInTheDocument();

    DEFAULT_CONTACT_LIST.forEach((contact) => {
      const contactName = screen.getByText(contact.name);
      expect(contactName).toBeInTheDocument();
    });
  });

  it("Renders the Searchable lists and all details with displayed email", () => {
    render(
      <SearchableList
        contacts={DEFAULT_CONTACT_LIST}
        onSelectedContactsChange={onSelectedContactsChange}
        displayEmail
      />
    );

    const searchInput = screen.getByPlaceholderText("Search");
    expect(searchInput).toBeInTheDocument();
    const attendedHeader = screen.getByText("Attended");
    expect(attendedHeader).toBeInTheDocument();
    const absentHeader = screen.getByText("Absent");
    expect(absentHeader).toBeInTheDocument();

    DEFAULT_CONTACT_LIST.forEach((contact) => {
      const contactName = screen.getByText(contact.name);
      expect(contactName).toBeInTheDocument();
    });
  });

  it("Triggers onSelectedContactsChange with the correct value when a contact is selected", () => {
    const selectedContacts: SelectedContactsRecord = {};
    render(
      <SearchableList
        contacts={DEFAULT_CONTACT_LIST}
        onSelectedContactsChange={onSelectedContactsChange}
      />
    );

    DEFAULT_CONTACT_LIST.forEach((contact) => {
      const contactName = screen.getByText(contact.name);
      act(() => {
        contactName.click();
      });

      selectedContacts[contact.id] = true;
      expect(onSelectedContactsChange).toHaveBeenCalledWith(selectedContacts);
    });

    // Make sure that all contacts were selected
    expect(Object.keys(selectedContacts).length).toBe(
      DEFAULT_CONTACT_LIST.length
    );
  });

  it("Filters the list when the search input is used", async () => {
    render(
      <SearchableList
        contacts={DEFAULT_CONTACT_LIST}
        onSelectedContactsChange={onSelectedContactsChange}
      />
    );

    const searchInput: HTMLInputElement = screen.getByPlaceholderText("Search");

    fireEvent.change(searchInput, {
      target: { value: DEFAULT_CONTACT_LIST[0].name },
    });

    const contactName = screen.getByText(DEFAULT_CONTACT_LIST[0].name);
    expect(contactName).toBeInTheDocument();
  });

  it("Filters the list with email when the search input is used and displayEmail is true", async () => {
    render(
      <SearchableList
        contacts={DEFAULT_CONTACT_LIST}
        onSelectedContactsChange={onSelectedContactsChange}
        displayEmail
      />
    );

    const searchInput: HTMLInputElement = screen.getByPlaceholderText("Search");

    fireEvent.change(searchInput, {
      target: { value: DEFAULT_CONTACT_LIST[0].email },
    });

    const contactName = screen.getByText(DEFAULT_CONTACT_LIST[0].name);
    expect(contactName).toBeInTheDocument();
  });

  it("Applies the correct selected styles when a contact is selected", () => {
    render(
      <SearchableList
        contacts={DEFAULT_CONTACT_LIST}
        onSelectedContactsChange={onSelectedContactsChange}
      />
    );
    const contactName = screen.getByText(DEFAULT_CONTACT_LIST[0].name);
    act(() => {
      contactName.click();
    });

    expect(contactName).toHaveStyleRule(
      `background-color: ${CONTENT_BACKGROUND_PRIMARY}`
    );
  });

  it("Defaults to the default image when none is provided", () => {
    render(
      <SearchableList
        contacts={[{ ...DEFAULT_CONTACT_LIST[0], displayImageURL: undefined }]}
        onSelectedContactsChange={onSelectedContactsChange}
      />
    );

    const displayImage = screen.getByAltText(
      `${DEFAULT_CONTACT_LIST[0].name}'s display image`
    );
    expect(displayImage).toHaveAttribute("src", DEFAULT_CONTACT_IMAGE_URL);
  });

  it("Should hide the list when the header is collapsed, and display it when it's expanded", () => {
    const [attended, absent] = DEFAULT_CONTACT_LIST.reduce<
      [Contact[], Contact[]]
    >(
      (acc, contact) => {
        if (contact.attended) {
          acc[0].push(contact);
        } else {
          acc[1].push(contact);
        }
        return acc;
      },
      [[], []]
    );

    render(
      <SearchableList
        contacts={DEFAULT_CONTACT_LIST}
        onSelectedContactsChange={onSelectedContactsChange}
      />
    );

    const attendedHeader = screen.getByText("Attended");
    const absentHeader = screen.getByText("Absent");

    expect(attendedHeader).toBeInTheDocument();
    expect(absentHeader).toBeInTheDocument();

    act(() => {
      attendedHeader.click();
    });

    attended.forEach((contact) => {
      const contactName = screen.queryByText(contact.name);
      expect(contactName).not.toBeInTheDocument();
    });

    act(() => {
      attendedHeader.click();
    });

    attended.forEach((contact) => {
      const contactName = screen.queryByText(contact.name);
      expect(contactName).toBeInTheDocument();
    });

    act(() => {
      absentHeader.click();
    });

    absent.forEach((contact) => {
      const contactName = screen.queryByText(contact.name);
      expect(contactName).not.toBeInTheDocument();
    });

    act(() => {
      absentHeader.click();
    });

    absent.forEach((contact) => {
      const contactName = screen.getByText(contact.name);
      expect(contactName).toBeInTheDocument();
    });
  });
});
