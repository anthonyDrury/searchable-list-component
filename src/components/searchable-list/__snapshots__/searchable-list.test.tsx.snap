// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchableList Matches the default snapshot 1`] = `
.c10 {
  height: 36px;
  width: 36px;
  border-radius: 50%;
}

.c11 {
  display: flex;
  flex-direction: column;
  justify-content: center;
}

.c12 {
  font-size: 14px;
  font-weight: 500;
  color: #20374B;
}

.c9 {
  cursor: pointer;
  display: flex;
  justify-content: flex-start;
  align-items: center;
  height: 52px;
  background-color: #FFFFFF;
  border-radius: 6px;
  padding: 8px;
  gap: 12px;
  max-width: 400px;
}

.c9:hover {
  background-color: #F2F5F7;
}

.c1 {
  background-color: #FFFFFF;
  display: flex;
  align-items: center;
  height: 48px;
  gap: 4px;
  font-size: 14px;
  color: #20374B;
  padding: 8px 16px;
  max-width: 400px;
}

.c2 {
  font-size: 14px;
  line-height: 20px;
  color: #20374B;
  border: none;
  outline: none;
}

.c2 textarea:focus,
.c2 input:focus {
  outline: none;
}

.c7 {
  display: flex;
  flex-direction: column;
  background-color: #FFFFFF;
}

.c5 {
  cursor: pointer;
  height: 37px;
  font-size: 14px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding-right: 8px;
  padding-left: 16px;
  background: #FFFFFF;
  border-bottom: 1px solid #F2F5F7;
  line-height: 20px;
  border-width: 1px 0px 1px 0px;
  border-style: solid;
  border-color: #E4E5E8;
  max-width: 400px;
}

.c3 {
  display: flex;
  flex-direction: column;
  max-width: 400px;
}

.c3:has(> .c4:hover) .c6,
.c3:has(> .c4:hover) .c8,
.c3:has(> .c4:hover) {
  background: #F2F5F7;
}

.c3:has(> .c4:hover) >.c4 {
  background: #F2F5F7;
}

.c0 {
  background-color: #FFFFFF;
  max-width: 400px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <svg
        fill="none"
        height="15"
        viewBox="0 0 15 15"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10.2589 9.66719H10.9172L14.4505 13.2172C14.7922 13.5589 14.7922 14.1172 14.4505 14.4589C14.1089 14.8005 13.5505 14.8005 13.2089 14.4589L9.66719 10.9172V10.2589L9.44219 10.0255C8.27552 11.0255 6.68385 11.5422 4.99219 11.2589C2.67552 10.8672 0.825521 8.93385 0.542188 6.60052C0.108855 3.07552 3.07552 0.108855 6.60052 0.542188C8.93385 0.825521 10.8672 2.67552 11.2589 4.99219C11.5422 6.68385 11.0255 8.27552 10.0255 9.44219L10.2589 9.66719ZM2.16719 5.91719C2.16719 7.99219 3.84219 9.66719 5.91719 9.66719C7.99219 9.66719 9.66719 7.99219 9.66719 5.91719C9.66719 3.84219 7.99219 2.16719 5.91719 2.16719C3.84219 2.16719 2.16719 3.84219 2.16719 5.91719Z"
          fill="#8E9AA5"
          fill-rule="evenodd"
        />
      </svg>
      <input
        class="c2"
        placeholder="Search"
      />
    </div>
    <div
      class="c3"
    >
      <div
        class="c4 c5"
      >
        <p>
          Attended
        </p>
        <svg
          fill="none"
          height="6"
          viewBox="0 0 6 10"
          width="10"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1.53243 0.259112L5.00447 3.69622L8.47651 0.259112C8.8255 -0.0863705 9.38926 -0.0863705 9.73825 0.259112C10.0872 0.604594 10.0872 1.16268 9.73825 1.50816L5.63087 5.57422C5.28187 5.9197 4.71811 5.9197 4.36912 5.57422L0.261732 1.50816C0.0941683 1.34266 0 1.11796 0 0.883637C0 0.649314 0.0941683 0.424617 0.261732 0.259112C0.610726 -0.077512 1.18343 -0.0863705 1.53243 0.259112Z"
            fill="#8E9AA5"
          />
        </svg>
      </div>
      <div
        class="c6 c7"
      >
        <div
          class="c8 c9"
        >
          <img
            alt="Alice Smith's display image"
            class="c10"
            src="/DISPLAY_IMAGE_2.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Alice Smith
            </p>
          </div>
        </div>
        <div
          class="c8 c9"
        >
          <img
            alt="Charlie Brown's display image"
            class="c10"
            src="/DISPLAY_IMAGE_4.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Charlie Brown
            </p>
          </div>
        </div>
        <div
          class="c8 c9"
        >
          <img
            alt="Eve Adams's display image"
            class="c10"
            src="/DISPLAY_IMAGE_6.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Eve Adams
            </p>
          </div>
        </div>
        <div
          class="c8 c9"
        >
          <img
            alt="Grace Hopper's display image"
            class="c10"
            src="/DISPLAY_IMAGE_8.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Grace Hopper
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4 c5"
      >
        <p>
          Absent
        </p>
        <svg
          fill="none"
          height="6"
          viewBox="0 0 6 10"
          width="10"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1.53243 0.259112L5.00447 3.69622L8.47651 0.259112C8.8255 -0.0863705 9.38926 -0.0863705 9.73825 0.259112C10.0872 0.604594 10.0872 1.16268 9.73825 1.50816L5.63087 5.57422C5.28187 5.9197 4.71811 5.9197 4.36912 5.57422L0.261732 1.50816C0.0941683 1.34266 0 1.11796 0 0.883637C0 0.649314 0.0941683 0.424617 0.261732 0.259112C0.610726 -0.077512 1.18343 -0.0863705 1.53243 0.259112Z"
            fill="#8E9AA5"
          />
        </svg>
      </div>
      <div
        class="c6 c7"
      >
        <div
          class="c8 c9"
        >
          <img
            alt="John Doe's display image"
            class="c10"
            src="/DISPLAY_IMAGE_0.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              John Doe
            </p>
          </div>
        </div>
        <div
          class="c8 c9"
        >
          <img
            alt="Jane Doe's display image"
            class="c10"
            src="/DISPLAY_IMAGE_1.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Jane Doe
            </p>
          </div>
        </div>
        <div
          class="c8 c9"
        >
          <img
            alt="Bob Johnson's display image"
            class="c10"
            src="/DISPLAY_IMAGE_3.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Bob Johnson
            </p>
          </div>
        </div>
        <div
          class="c8 c9"
        >
          <img
            alt="Diana Prince's display image"
            class="c10"
            src="/DISPLAY_IMAGE_5.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Diana Prince
            </p>
          </div>
        </div>
        <div
          class="c8 c9"
        >
          <img
            alt="Frank Castle's display image"
            class="c10"
            src="/DISPLAY_IMAGE_7.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Frank Castle
            </p>
          </div>
        </div>
        <div
          class="c8 c9"
        >
          <img
            alt="Hank Pym's display image"
            class="c10"
            src="/DISPLAY_IMAGE_9.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Hank Pym
            </p>
          </div>
        </div>
        <div
          class="c8 c9"
        >
          <img
            alt="Ivy Wayne's display image"
            class="c10"
            src="/default_contact_display_image.jpg"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Ivy Wayne
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SearchableList Matches the email display snapshot 1`] = `
.c10 {
  height: 36px;
  width: 36px;
  border-radius: 50%;
}

.c11 {
  display: flex;
  flex-direction: column;
  justify-content: center;
}

.c12 {
  font-size: 14px;
  font-weight: 500;
  color: #20374B;
}

.c13 {
  font-size: 13px;
  color: #5A6D80;
}

.c9 {
  cursor: pointer;
  display: flex;
  justify-content: flex-start;
  align-items: center;
  height: 52px;
  background-color: #FFFFFF;
  border-radius: 6px;
  padding: 8px;
  gap: 12px;
  max-width: 400px;
}

.c9:hover {
  background-color: #F2F5F7;
}

.c1 {
  background-color: #FFFFFF;
  display: flex;
  align-items: center;
  height: 48px;
  gap: 4px;
  font-size: 14px;
  color: #20374B;
  padding: 8px 16px;
  max-width: 400px;
}

.c2 {
  font-size: 14px;
  line-height: 20px;
  color: #20374B;
  border: none;
  outline: none;
}

.c2 textarea:focus,
.c2 input:focus {
  outline: none;
}

.c7 {
  display: flex;
  flex-direction: column;
  background-color: #FFFFFF;
}

.c5 {
  cursor: pointer;
  height: 37px;
  font-size: 14px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding-right: 8px;
  padding-left: 16px;
  background: #FFFFFF;
  border-bottom: 1px solid #F2F5F7;
  line-height: 20px;
  border-width: 1px 0px 1px 0px;
  border-style: solid;
  border-color: #E4E5E8;
  max-width: 400px;
}

.c3 {
  display: flex;
  flex-direction: column;
  max-width: 400px;
}

.c3:has(> .c4:hover) .c6,
.c3:has(> .c4:hover) .c8,
.c3:has(> .c4:hover) {
  background: #F2F5F7;
}

.c3:has(> .c4:hover) >.c4 {
  background: #F2F5F7;
}

.c0 {
  background-color: #FFFFFF;
  max-width: 400px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <svg
        fill="none"
        height="15"
        viewBox="0 0 15 15"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10.2589 9.66719H10.9172L14.4505 13.2172C14.7922 13.5589 14.7922 14.1172 14.4505 14.4589C14.1089 14.8005 13.5505 14.8005 13.2089 14.4589L9.66719 10.9172V10.2589L9.44219 10.0255C8.27552 11.0255 6.68385 11.5422 4.99219 11.2589C2.67552 10.8672 0.825521 8.93385 0.542188 6.60052C0.108855 3.07552 3.07552 0.108855 6.60052 0.542188C8.93385 0.825521 10.8672 2.67552 11.2589 4.99219C11.5422 6.68385 11.0255 8.27552 10.0255 9.44219L10.2589 9.66719ZM2.16719 5.91719C2.16719 7.99219 3.84219 9.66719 5.91719 9.66719C7.99219 9.66719 9.66719 7.99219 9.66719 5.91719C9.66719 3.84219 7.99219 2.16719 5.91719 2.16719C3.84219 2.16719 2.16719 3.84219 2.16719 5.91719Z"
          fill="#8E9AA5"
          fill-rule="evenodd"
        />
      </svg>
      <input
        class="c2"
        placeholder="Search"
      />
    </div>
    <div
      class="c3"
    >
      <div
        class="c4 c5"
      >
        <p>
          Attended
        </p>
        <svg
          fill="none"
          height="6"
          viewBox="0 0 6 10"
          width="10"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1.53243 0.259112L5.00447 3.69622L8.47651 0.259112C8.8255 -0.0863705 9.38926 -0.0863705 9.73825 0.259112C10.0872 0.604594 10.0872 1.16268 9.73825 1.50816L5.63087 5.57422C5.28187 5.9197 4.71811 5.9197 4.36912 5.57422L0.261732 1.50816C0.0941683 1.34266 0 1.11796 0 0.883637C0 0.649314 0.0941683 0.424617 0.261732 0.259112C0.610726 -0.077512 1.18343 -0.0863705 1.53243 0.259112Z"
            fill="#8E9AA5"
          />
        </svg>
      </div>
      <div
        class="c6 c7"
      >
        <div
          class="c8 c9"
        >
          <img
            alt="Alice Smith's display image"
            class="c10"
            src="/DISPLAY_IMAGE_2.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Alice Smith
            </p>
            <p
              class="c13"
            >
              alice@example.com
            </p>
          </div>
        </div>
        <div
          class="c8 c9"
        >
          <img
            alt="Charlie Brown's display image"
            class="c10"
            src="/DISPLAY_IMAGE_4.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Charlie Brown
            </p>
            <p
              class="c13"
            >
              charlie@example.com
            </p>
          </div>
        </div>
        <div
          class="c8 c9"
        >
          <img
            alt="Eve Adams's display image"
            class="c10"
            src="/DISPLAY_IMAGE_6.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Eve Adams
            </p>
            <p
              class="c13"
            >
              eve@example.com
            </p>
          </div>
        </div>
        <div
          class="c8 c9"
        >
          <img
            alt="Grace Hopper's display image"
            class="c10"
            src="/DISPLAY_IMAGE_8.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Grace Hopper
            </p>
            <p
              class="c13"
            >
              grace@example.com
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4 c5"
      >
        <p>
          Absent
        </p>
        <svg
          fill="none"
          height="6"
          viewBox="0 0 6 10"
          width="10"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1.53243 0.259112L5.00447 3.69622L8.47651 0.259112C8.8255 -0.0863705 9.38926 -0.0863705 9.73825 0.259112C10.0872 0.604594 10.0872 1.16268 9.73825 1.50816L5.63087 5.57422C5.28187 5.9197 4.71811 5.9197 4.36912 5.57422L0.261732 1.50816C0.0941683 1.34266 0 1.11796 0 0.883637C0 0.649314 0.0941683 0.424617 0.261732 0.259112C0.610726 -0.077512 1.18343 -0.0863705 1.53243 0.259112Z"
            fill="#8E9AA5"
          />
        </svg>
      </div>
      <div
        class="c6 c7"
      >
        <div
          class="c8 c9"
        >
          <img
            alt="John Doe's display image"
            class="c10"
            src="/DISPLAY_IMAGE_0.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              John Doe
            </p>
          </div>
        </div>
        <div
          class="c8 c9"
        >
          <img
            alt="Jane Doe's display image"
            class="c10"
            src="/DISPLAY_IMAGE_1.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Jane Doe
            </p>
            <p
              class="c13"
            >
              test@test.com
            </p>
          </div>
        </div>
        <div
          class="c8 c9"
        >
          <img
            alt="Bob Johnson's display image"
            class="c10"
            src="/DISPLAY_IMAGE_3.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Bob Johnson
            </p>
            <p
              class="c13"
            >
              bob@example.com
            </p>
          </div>
        </div>
        <div
          class="c8 c9"
        >
          <img
            alt="Diana Prince's display image"
            class="c10"
            src="/DISPLAY_IMAGE_5.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Diana Prince
            </p>
            <p
              class="c13"
            >
              diana@example.com
            </p>
          </div>
        </div>
        <div
          class="c8 c9"
        >
          <img
            alt="Frank Castle's display image"
            class="c10"
            src="/DISPLAY_IMAGE_7.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Frank Castle
            </p>
            <p
              class="c13"
            >
              frank@example.com
            </p>
          </div>
        </div>
        <div
          class="c8 c9"
        >
          <img
            alt="Hank Pym's display image"
            class="c10"
            src="/DISPLAY_IMAGE_9.png"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Hank Pym
            </p>
            <p
              class="c13"
            >
              hank@example.com
            </p>
          </div>
        </div>
        <div
          class="c8 c9"
        >
          <img
            alt="Ivy Wayne's display image"
            class="c10"
            src="/default_contact_display_image.jpg"
          />
          <div
            class="c11"
          >
            <p
              class="c12"
            >
              Ivy Wayne
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
